File: src/main/java/com/fifo/ticketing/global/config/AsyncConfig.java
package com.fifo.ticketing.global.config;

import java.util.concurrent.Executor;
import java.util.concurrent.ThreadPoolExecutor;
import org.springframework.aop.interceptor.AsyncUncaughtExceptionHandler;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.scheduling.annotation.AsyncConfigurer;
import org.springframework.scheduling.annotation.EnableAsync;
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor;

@EnableAsync
@Configuration
public class AsyncConfig implements AsyncConfigurer {


    //커스텀 쓰레드 풀 정의
    @Override
    @Bean(name = "mailExecutor")
    public Executor getAsyncExecutor() {
        ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
        executor.setCorePoolSize(5);
        executor.setMaxPoolSize(10);
        executor.setQueueCapacity(100);
        executor.initialize();

        return executor;
    }

    @Bean(name = "cancelPerformanceMailExecutor")
    public Executor cancelMailExecutor() {
        ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();

        executor.setCorePoolSize(4);
        executor.setMaxPoolSize(10);
        executor.setQueueCapacity(50);
        executor.setKeepAliveSeconds(60);
        executor.setThreadNamePrefix("async-exec-mail-");
        executor.setRejectedExecutionHandler(new ThreadPoolExecutor.CallerRunsPolicy());
        executor.initialize();
        return executor;
    }


}

File: src/main/java/com/fifo/ticketing/global/config/AsyncConfig.kt
package com.fifo.ticketing.global.config

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.scheduling.annotation.AsyncConfigurer
import org.springframework.scheduling.annotation.EnableAsync
import java.util.concurrent.Executor
import java.util.concurrent.Executors

@EnableAsync
@Configuration
class AsyncConfig : AsyncConfigurer {

    @Bean(name = ["mailExecutor"])
    override fun getAsyncExecutor(): Executor =
        Executors.newVirtualThreadPerTaskExecutor()

    @Bean(name = ["cancelPerformanceMailExecutor"])
    fun cancelMailExecutor(): Executor =
        Executors.newVirtualThreadPerTaskExecutor()
}
